contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1.sol

```diff
@@ -0,0 +1,1646 @@
+        _split,
+        _projectId,
+        _group,
+        _payoutAmount,
+        _feeDiscount
+      );
+
+      // If the split allocator is set as feeless, this distribution is not eligible for a fee.
+      if (_netPayoutAmount != 0 && _netPayoutAmount != _payoutAmount)
+        feeEligibleDistributionAmount += _payoutAmount;
+
+      if (_payoutAmount > 0) {
+        // Subtract from the amount to be sent to the beneficiary.
+        unchecked {
+          leftoverAmount = leftoverAmount - _payoutAmount;
+        }
+      }
+
+      emit DistributeToPayoutSplit(
+        _projectId,
+        _domain,
+        _group,
+        _split,
+        _netPayoutAmount,
+        msg.sender
+      );
+
+      unchecked {
+        ++_i;
+      }
+    }
+  }
+
+  /**
+    @notice
+    Pays out a split for a project's funding cycle configuration.
+  
+    @param _split The split to distribute payouts to.
+    @param _amount The total amount being distributed to the split, as a fixed point number with the same number of decimals as this terminal.
+    @param _feeDiscount The amount of discount to apply to the fee, out of the MAX_FEE.
+
+    @return netPayoutAmount The amount sent to the split after subtracting fees.
+  */
+  function _distributeToPayoutSplitsOf(
+    JBSplit memory _split,
+    uint256 _projectId,
+    uint256 _group,
+    uint256 _amount,
+    uint256 _feeDiscount
+  ) internal returns (uint256 netPayoutAmount) {
+    // If there's an allocator set, transfer to its `allocate` function.
+    if (_split.allocator != IJBSplitAllocator(address(0))) {
+      // If the split allocator is set as feeless, this distribution is not eligible for a fee.
+      if (
+        _feeDiscount == JBConstants.MAX_FEE_DISCOUNT || isFeelessAddress[address(_split.allocator)]
+      )
+        netPayoutAmount = _amount;
+        // This distribution is eligible for a fee since the funds are leaving this contract and the allocator isn't listed as feeless.
+      else {
+        unchecked {
+          netPayoutAmount = _amount - _feeAmount(_amount, fee, _feeDiscount);
+        }
+      }
+
+      // Trigger any inherited pre-transfer logic.
+      _beforeTransferTo(address(_split.allocator), netPayoutAmount);
+
+      // Create the data to send to the allocator.
+      JBSplitAllocationData memory _data = JBSplitAllocationData(
+        token,
+        netPayoutAmount,
+        decimals,
+        _projectId,
+        _group,
+        _split
+      );
+
+      // Trigger the allocator's `allocate` function.
+      // If this terminal's token is ETH, send it in msg.value.
+      try
+        _split.allocator.allocate{value: token == JBTokens.ETH ? netPayoutAmount : 0}(_data)
+      {} catch {
+        // Set the net payout amount to 0 to signal the reversion.
+        netPayoutAmount = 0;
+
+        // Add undistributed amount back to project's balance.
+        store.recordAddedBalanceFor(_projectId, _amount);
```

Comment:
In ERC20 terminals this is exploitable a malicious user can:
1. Create a project
2. Deposit some amount of an ERC20
3. Create an Allocator that always reverts when `allocate` gets called
4. Distribute all funds to the allocator
5. The terminal calls `_beforeTransferTo` which sets an allowance, but the call reverts, the balance gets added back to the project
6. The malicious user calls a method on their allocator to use the allowance and transfer it to them
7. Go back to step 4

We could change `_beforeTransferTo` to increase the allowance and add another method (ex. `_cancelTransferTo`) to decrease it, in the catch we then decrease it with `_amount` to get back to the original allowance.

Reply:
yes. agree with fix. worth doing.

Reply:
nice catch